plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/DataSourceSerializerModern.java

```diff
@@ -796,6 +775,89 @@ public boolean parseDataSources(
 
     }
 
+    @Nullable
+    private Map<String, Map<String, Map<String, String>>> readSecureCredentials(
+        @NotNull DBPDataSourceConfigurationStorage configurationStorage,
+        @NotNull DataSourceConfigurationManager configurationManager,
+        @Nullable Collection<String> dataSourceIds
+    ) throws DBException {
+        if (configurationManager.isSecure()) {
+            return null;
+        }
+        final String name = "%s%s%s".formatted(
+            DBPDataSourceRegistry.CREDENTIALS_CONFIG_FILE_PREFIX,
+            configurationStorage.getStorageSubId(),
+            DBPDataSourceRegistry.CREDENTIALS_CONFIG_FILE_EXT
+        );
+        try (InputStream is = configurationManager.readConfiguration(name, dataSourceIds)) {
+            if (is == null) {
+                return null;
+            }
+            final String data = loadConfigFile(is, true);
+            return CONFIG_GSON.fromJson(data, new TypeToken<Map<String, Map<String, Map<String, String>>>>() {}.getType());
+        } catch (Exception e) {
+            if (!DBWorkbench.getPlatform().getApplication().isHeadlessMode()) {
+                if (DBWorkbench.getPlatformUI().confirmAction(
+                    RegistryMessages.project_open_cannot_read_credentials_title,
+                    NLS.bind(
+                        RegistryMessages.project_open_cannot_read_credentials_message,
+                        registry.getProject().getName()
+                    ),
+                    RegistryMessages.project_open_cannot_read_credentials_button_text,
+                    true
+                )) {
+                    final String backupName = name + ".bak";
+                    try (InputStream is = configurationManager.readConfiguration(backupName, dataSourceIds)) {
+                        if (is != null) { // just sanity check
+                            configurationManager.writeConfiguration(backupName, is.readAllBytes());
+                        }
+                        log.debug("A backup file was created that contains secure credentials: " + backupName);
+                    } catch (Exception e1) {
+                        log.error("Error backing up secure credentials", e1);
+                    }
+                } else {
+                    throw new DBException("Project opening canceled by user");
+                }
+            }
+            log.error("Error reading secure credentials", e);
+        }
+        return null;
+    }
+
+    @Nullable
+    private Map<String, Object> readConfiguration(
+        @NotNull DBPDataSourceConfigurationStorage configurationStorage,
+        @NotNull DataSourceConfigurationManager configurationManager,
+        @Nullable Collection<String> dataSourceIds
+    ) throws DBException, IOException {
+        final InputStream is;
```

Comment:
Input stream is never closed

Reply:
It's closed inside the `IOUtils#copyStream` method below.